File: client/src/lib/levelGenerators/tetris.ts
/**
 * Tetris Engine & Polyomino Tables
 * Original bitwise 16-bit masks from jakesgordon/javascript-tetris (Jake Gordon, MIT License)
 * Adapted to SELY MiniGame Hub modular architecture.
 */

export interface TetrisPieceType {
  name: string;
  size: number;
  blocks: number[]; // 4 rotation states represented as 16-bit integer masks
  color: string;
}

export interface ActivePiece {
  type: TetrisPieceType;
  dir: number; // 0..3
  x: number;
  y: number;
}

export const TETRIS_PIECES: Record<string, TetrisPieceType> = {
  i: { name: "I", size: 4, blocks: [0x0f00, 0x2222, 0x00f0, 0x4444], color: "#293B75" },
  j: { name: "J", size: 3, blocks: [0x44c0, 0x8e00, 0x6440, 0x0e20], color: "#296A55" },
  l: { name: "L", size: 3, blocks: [0x4460, 0x0e80, 0xc440, 0x2e00], color: "#E5B341" },
  o: { name: "O", size: 2, blocks: [0xcc00, 0xcc00, 0xcc00, 0xcc00], color: "#E9563F" },
  s: { name: "S", size: 3, blocks: [0x06c0, 0x8c40, 0x6c00, 0x4620], color: "#654169" },
  t: { name: "T", size: 3, blocks: [0x0e40, 0x4c40, 0x4e00, 0x4640], color: "#1B1A1B" },
  z: { name: "Z", size: 3, blocks: [0x0c60, 0x4c80, 0xc600, 0x2640], color: "#D97706" },
};

export const TETRIS_CONFIG = {
  nx: 10, // Width of court (blocks)
  ny: 20, // Height of court (blocks)
  speedStart: 0.6, // Seconds before drop
  speedDecrement: 0.006,
  speedMin: 0.09,
};

/**
 * Iterates through each block of a piece rotation mask
 */
export function forEachPieceBlock(
  type: TetrisPieceType,
  x: number,
  y: number,
  dir: number,
  fn: (bx: number, by: number) => void
) {
  let row = 0;
  let col = 0;
  const blocks = type.blocks[dir];
  for (let bit = 0x8000; bit > 0; bit = bit >> 1) {
    if (blocks & bit) {
      fn(x + col, y + row);
    }
    if (++col === 4) {
      col = 0;
      ++row;
    }
  }
}

/**
 * Checks if a piece collides with court boundaries or placed blocks
 */
export function isPieceOccupied(
  type: TetrisPieceType,
  x: number,
  y: number,
  dir: number,
  court: (string | null)[][]
): boolean {
  let result = false;
  forEachPieceBlock(type, x, y, dir, (bx, by) => {
    if (bx < 0 || bx >= TETRIS_CONFIG.nx || by < 0 || by >= TETRIS_CONFIG.ny || court[by]?.[bx] != null) {
      result = true;
    }
  });
  return result;
}

/**
 * 7-Bag Randomizer
 */
export function createTetrisBag(seed: number): () => TetrisPieceType {
  let bag: TetrisPieceType[] = [];
  const pieces = Object.values(TETRIS_PIECES);

  let rngVal = (seed ^ 0x9e3779b9) >>> 0;
  const nextRng = () => {
    rngVal = (Math.imul(rngVal, 1664525) + 1013904223) >>> 0;
    return rngVal / 4294967296;
  };

  return function nextPiece(): TetrisPieceType {
    if (bag.length === 0) {
      bag = [...pieces];
      // Fisher-Yates shuffle with deterministic RNG
      for (let i = bag.length - 1; i > 0; i--) {
        const j = Math.floor(nextRng() * (i + 1));
        [bag[i], bag[j]] = [bag[j], bag[i]];
      }
    }
    return bag.pop()!;
  };
}

File: src/games/tetris.rs
//! Tetris Game Engine & Bitwise Collision Logic
//! Ported from jakesgordon/javascript-tetris.
//! 16-bit hex piece bitmasks, 10x20 grid, 7-bag randomizer and bitwise collision tests.

use serde::{Deserialize, Serialize};

pub const TETRIS_COLS: usize = 10;
pub const TETRIS_ROWS: usize = 20;

/// 16-bit piece definitions in 4 orientations (0°, 90°, 180°, 270°)
pub const PIECE_I: [u16; 4] = [0x0F00, 0x2222, 0x00F0, 0x4444];
pub const PIECE_J: [u16; 4] = [0x44C0, 0x8E00, 0x6440, 0x0E20];
pub const PIECE_L: [u16; 4] = [0x4460, 0x0E80, 0xC440, 0x2E00];
pub const PIECE_O: [u16; 4] = [0xCC00, 0xCC00, 0xCC00, 0xCC00];
pub const PIECE_S: [u16; 4] = [0x06C0, 0x8C40, 0x6C00, 0x4620];
pub const PIECE_T: [u16; 4] = [0x0E40, 0x4C40, 0x4E00, 0x4640];
pub const PIECE_Z: [u16; 4] = [0x0C60, 0x4C80, 0xC600, 0x2640];

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum TetrominoType {
    I,
    J,
    L,
    O,
    S,
    T,
    Z,
}

impl TetrominoType {
    pub fn get_rotations(&self) -> [u16; 4] {
        match self {
            TetrominoType::I => PIECE_I,
            TetrominoType::J => PIECE_J,
            TetrominoType::L => PIECE_L,
            TetrominoType::O => PIECE_O,
            TetrominoType::S => PIECE_S,
            TetrominoType::T => PIECE_T,
            TetrominoType::Z => PIECE_Z,
        }
    }

    pub fn color_hex(&self) -> &'static str {
        match self {
            TetrominoType::I => "#296A55",
            TetrominoType::J => "#293B75",
            TetrominoType::L => "#E5B341",
            TetrominoType::O => "#654169",
            TetrominoType::S => "#E9563F",
            TetrominoType::T => "#1B1A1B",
            TetrominoType::Z => "#D9381E",
        }
    }
}

/// Checks bitwise collision between a piece at (x, y) with rotation and a 10x20 grid
pub fn check_tetris_collision(
    piece_type: TetrominoType,
    rot_idx: usize,
    x: i32,
    y: i32,
    grid: &[[Option<&'static str>; TETRIS_COLS]; TETRIS_ROWS],
) -> bool {
    let mask = piece_type.get_rotations()[rot_idx % 4];
    for bit in 0..16 {
        if (mask & (0x8000 >> bit)) != 0 {
            let col = x + (bit % 4);
            let row = y + (bit / 4);

            if col < 0 || col >= (TETRIS_COLS as i32) || row >= (TETRIS_ROWS as i32) {
                return true; // Wall or floor collision
            }
            if row >= 0 && grid[row as usize][col as usize].is_some() {
                return true; // Occupied block collision
            }
        }
    }
    false
}
